File: handlers/hostinfo/memory.go
package hostInfo

import (
	"bufio"
	"github.com/Sirupsen/logrus"
	"math"
	"os"
	"runtime"
	"strconv"
	"strings"
)

var KeyMap = map[string]string{
	"memtotal":     "memTotal",
	"memfree":      "memFree",
	"memavailable": "memAvailable",
	"buffers":      "buffers",
	"cached":       "cached",
	"swapcached":   "swapCached",
	"active":       "active",
	"inactive":     "inactive",
	"swaptotal":    "swapTotal",
	"swapfree":     "swapFree",
}

type MemoryCollector struct {
	unit float64
}

func (m MemoryCollector) getMemInfoData() []string {
	file, err := os.Open("/proc/meminfo")
	defer file.Close()
	data := []string{}
	if err != nil {
		logrus.Error(err)
	} else {
		scanner := bufio.NewScanner(file)
		scanner.Split(bufio.ScanLines)
		for scanner.Scan() {
			data = append(data, scanner.Text())
		}
	}
	logrus.Infof("memory info %v", data)
	return data
}

func (m MemoryCollector) parseLinuxMemInfo() map[string]interface{} {
	data := map[string]interface{}{}
	memData := m.getMemInfoData()
	for _, line := range memData {
		lineList := strings.Split(line, ":")
		keyLower := strings.ToLower(lineList[0])
		possibleMemValue := strings.Split(strings.TrimSpace(lineList[1]), " ")[0]

		if _, ok := KeyMap[keyLower]; ok {
			value, _ := strconv.ParseFloat(possibleMemValue, 64)
			convertMemVal := value / m.unit
			convertMemVal = math.Floor(convertMemVal*1000) / 1000
			data[KeyMap[keyLower]] = convertMemVal
		}
	}
	return data
}

func (m MemoryCollector) GetData() map[string]interface{} {
	if runtime.GOOS == "linux" {
		return m.parseLinuxMemInfo()
	}
	return map[string]interface{}{}
}

func (m MemoryCollector) KeyName() string {
	return "memoryInfo"
}

func (m MemoryCollector) GetLabels(prefix string) map[string]string {
	return map[string]string{}
}

File: tests/tests/cattle/plugins/host_info/memory.py
import platform


class MemoryCollector(object):
    def __init__(self):
        self.key_map = {'memtotal': 'memTotal',
                        'memfree': 'memFree',
                        'memavailable': 'memAvailable',
                        'buffers': 'buffers',
                        'cached': 'cached',
                        'swapcached': 'swapCached',
                        'active': 'active',
                        'inactive': 'inactive',
                        'swaptotal': 'swapTotal',
                        'swapfree': 'swapFree'
                        }

        self.unit = 1024.0

    def _get_meminfo_data(self):
        with open('/proc/meminfo') as f:
            return f.readlines()

    def _parse_linux_meminfo(self):
        data = {k: None for k in self.key_map.values()}

        # /proc/meminfo file has all values in kB
        mem_data = self._get_meminfo_data()
        for line in mem_data:
            line_list = line.split(':')
            key_lower = line_list[0].lower()
            possible_mem_value = line_list[1].strip().split(' ')[0]

            if self.key_map.get(key_lower):
                converted_mem_val = float(possible_mem_value)/self.unit
                data[self.key_map[key_lower]] = round(converted_mem_val, 3)

        return data

    def key_name(self):
        return "memoryInfo"

    def get_data(self):
        if platform.system() == 'Linux':
            return self._parse_linux_meminfo()
        else:
            return {}
